src/tools/clippy/clippy_utils/src/ast_utils.rs

```diff
@@ -143,7 +143,7 @@ pub fn eq_expr(l: &Expr, r: &Expr) -> bool {
     match (&l.kind, &r.kind) {
         (Paren(l), _) => eq_expr(l, r),
         (_, Paren(r)) => eq_expr(l, r),
-        (Err, Err) => true,
+        (Err(_), Err(_)) => true,
```

Comment:
`(Dummy, Dummy)` is missing from this match statement.

Reply:
To mimic the old behavior `(Err(_), Dummy)` and `(Dummy, Err(_))` may also be needed, and I wonder if `(Dummy, _) | (_, Dummy) => true` makes sense.

But the only two occurences of `ExprKind::Dummy` are in
- `<Expr as DummyAstNode>::dummy()` which isn't actually called (I can replace its body with `panic!()` and both `./x test --stage 1 tests/ui/` and `./x check --stage 2` passes successfully)
  https://github.com/rust-lang/rust/blob/8c0b4f67c9b4bf477df38c16669fd576d46a2b3d/compiler/rustc_ast/src/mut_visit.rs#L1635-L1645
- `Parser::maybe_recover_trailing_expr()` for avoiding creating a `P<Expr>`
  https://github.com/rust-lang/rust/blob/8c0b4f67c9b4bf477df38c16669fd576d46a2b3d/compiler/rustc_parse/src/parser/pat.rs#L388-L394

So `ast_utils.rs`'s `eq_expr` cannot actually have an `ExprKind::Dummy`, so this case can't be tested in the testsuite. I was tempted to put a `panic!()` to hold the choice until later if someone uses `Dummy` and it reaches Clippy, but if it's not tested I'd rather have a wrong diagnostic than a panic in prod.

Reply:
If this can really not occur here, then I suggest to add a `debug_assertion` in this function. I'd like this to be an exhaustive match.